### ISSUE TITLE
Stepping in the compiler doesn't work

### ISSUE THREAD

**Issue by lucaregini** [gambit/gambit#648]
Steps to reproduce:

- load igsc.scm in the interpreter -> gsi/gsi -:=. gsc/igsc.scm
-set a breakpoint in compile-program -> (break c#compile-program)
-compile any sample program -> (compile-file "h.scm")   
-The code will break at compile-program: -> *** STOPPED IN c#compile-program, "gsc/_front.scm"@299.29
- Try to step by entering ,s in the REPL.

This results in a deadlock:

 * ERROR IN #<procedure #2>, "_front.scm"@68.22 -- Deadlock detected"

**Comment by feeley** [comment]
The deadlock error is due to a combination of things but the root cause is the following definition in `gsc/_gscdebug.scm` (which is the last file loaded by `gsc/igsc.scm`):

```
(set! ##wr (lambda (we obj) (##default-wr we ...)))
```

This extends the Gambit `##wr` procedure (used to implement `display`, `write`, `pretty-print`, ...) so that it prints `obj` nicely when it is a circular object created by the compiler (such as a parse tree).

Because `gsc/_gscdebug.scm` is loaded by `gsc/igsc.scm` the procedure in `##wr` is an interpreted procedure and like any other interpreted procedure its code will be single stepped if the `,s` command is used at the REPL.

The problem is that the single-stepper needs to generate output to show what is being single-stepped, and you guessed it, this means calling `##wr`.  But this will also want to generate output by calling `##wr`.  The runtime system detects this circularity and gives a deadlock exception.

A simple way to avoid the problem is to comment out the filename `_gscdebug.scm` in `gsc/igsc.scm`.  This avoids the circularity as shown in the trace below.

This is not ideal but should allow you to keep working in the short term.  I will look into a more convenient approach for more intense debugging of the compiler.


```
% gsi/gsi -:=. gsc/igsc.scm
loading "~~/gsc/_host.scm"
...
loading "~~/gsc/_t-c-1.scm"
loading "~~/gsc/_t-c-3.scm"
loading "~~/gsc/_t-c-2.scm"
loading "~~/gsc/_gsclib.scm"
loading "~~/gsc/_gsc.scm"
Gambit v4.9.3-1234-g6acd87df

> (begin (break c#normalize-program) (compile-file "h.scm"))
*** STOPPED IN c#normalize-program, "gsc/_ptree2.scm"@29.15
1> ,e
ptrees = '(#((app-tag) #f (#((ref-tag) #((app-tag) #f (#((ref-tag) #((app... #2
1> (load "gsc/_gscdebug.scm")
"/Users/feeley/gambit/gsc/_gscdebug.scm"
1> ,e
ptrees = '((PTREE: (println "hello")))
1> (set! ##wr ##default-wr)
1> ,s
| > ptrees
| (#((app-tag) #f (#((ref-tag) #((app-tag) #f (#((ref-tag) #((app-tag) #f (#...
*** STOPPED IN c#normalize-program, "gsc/_ptree2.scm"@30.16
1> ,s
| > c#expand-primitive-calls
| #<procedure #3 c#expand-primitive-calls>
*** STOPPED IN c#normalize-program, "gsc/_ptree2.scm"@30.39
1> ,s
| > lst
| (#((app-tag) #f (#((ref-tag) #((app-tag) #f (#((ref-tag) #((app-tag) #f (#...
*** STOPPED IN c#normalize-program, "gsc/_ptree2.scm"@30.15
1> ,s
| > (c#expand-primitive-calls lst)
*** STOPPED IN c#expand-primitive-calls, "gsc/_ptree2.scm"@163.4
1> 
```

**Comment by feeley** [comment]
It turns out the best approach was to adjust the single stepper so that it allows stepping even when `##wr` is set to an interpreted procedure.  Commit 85614c85934afd275cca697128da77bda14ffee1 is the fix.  Here is a trace of the compiler with the new stepper:

**Comment by feeley** [comment]
Let me add that after some more digging the source of the deadlock is now clear and is related to the mutex that controls access to the console (tty) when a REPL is started (to allow multiple Scheme threads to operate their own REPL on a shared console).  The problem is that the single-stepping and tracing messages are written to the console in a critical section that has acquired the "console mutex".  However, when `##wr` is redefined to an interpreted procedure (by `gsc/_gscdebug.scm`) and it is single stepped or traced (or one of the procedure it calls such as `c#var-name`, `list`, etc is single stepped or traced) then the console mutex must again be acquired to output these messages.  This leads to a deadlock because the mutex is not free to be acquired.  This has been fixed (commit a102b4d5bf98f826c00f2a52c619885ab767c3f0) by generating the message in a string before the console mutex is acquired.
